### ISSUE TITLE
plugins.okru: Could not find metadata and cannot record stream

### ISSUE THREAD

**Issue by akali752** [streamlink/streamlink#4414]
### Checklist

- [X] This is a plugin issue and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin issues](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22plugin+issue%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest stable release

### Description

Recently when I tried to record ok.ru live stream, it always shows "[plugins.okru][error] Could not find metadata" (it was prefectly fine and no issue like 2 months ago). The stream is live and there is playable streams in the URL, and I also can get m3u8 URL by using youtube-dl.

```
streamlink "https://ok.ru/video/3178148273793"
[cli][info] Found matching plugin okru for URL https://ok.ru/video/3178148273793
[plugins.okru][error] Could not find metadata
error: No playable streams found on this URL: https://ok.ru/video/3178148273793
```

### Debug log

```text
streamlink "https://ok.ru/video/3178148273793"  --loglevel debug
[cli][debug] OS:         macOS 10.13.6
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.2.0
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.3.1)
[cli][debug] Arguments:
[cli][debug]  url=https://ok.ru/video/3178148273793
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin okru for URL https://ok.ru/video/3178148273793
[plugins.okru][error] Could not find metadata
error: No playable streams found on this URL: https://ok.ru/video/3178148273793
```

**Comment by bastimeyer** [comment]
> [plugins.okru][error] Could not find metadata

I can't reproduce this.

The site doesn't appear to have changed anything and Streamlink's okru plugin does successfully find the matadata required for getting the HLS stream URL, as you can see here:

```
$ streamlink -l debug 'https://ok.ru/video/3178148273793'
[cli][debug] OS:         Linux-5.17.1-1-git-x86_64-with-glibc2.35
[cli][debug] Python:     3.10.4
[cli][debug] Streamlink: 3.2.0+9.gfeec5f4
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.3.1)
[cli][debug] Arguments:
[cli][debug]  url=https://ok.ru/video/3178148273793
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][info] Found matching plugin okru for URL https://ok.ru/video/3178148273793
[utils.l10n][debug] Language code: en_US
Available streams: 240p (worst), 360p, 480p, 720p (best)
```

https://github.com/streamlink/streamlink/blame/3.2.0/src/streamlink/plugins/okru.py#L63-L81

This is essentially what the plugin does, which you can check yourself in your web browser's dev tools:

```js
JSON.parse(JSON.parse(document.querySelector("[data-options]").dataset.options).flashvars.metadata)
```

**Comment by akali752** [comment]
Thanks very much, thisisgk! You remind me that this issue did start to happen after I upgraded streamlink, I have not upgraded streamlink for a while before that, so the previous version that I was using should be 3.0.1 or lower.

**Comment by akali752** [comment]
> Sorry to interrupt.
> 
> This issue may be related to a specific macOS version, but may be "fixed" or "worked around" with streamlink.
> 
> I have tried this, and while it doesn't reproduce on macOS 10.15.7, it reproduces on 10.12.6 (both with Python 3.10.4 `venv`, then fresh `pip install streamlink`). I also tried older streamlink versions on macOS 10.12.6, and it reproduces with 3.0.2 or above, but not with 3.0.1 or 2.4.0.
> 
> I modified `parse_html()` closer to 3.0.1, then it no longer reproduces (relating commit [12c17c4](https://github.com/streamlink/streamlink/commit/12c17c4ffb1a4236eb183513ca7805718fbc1ab2)).
> 
> ```diff
> --- ./streamlink/utils/parse.py.orig	2022-03-25 09:19:39.000000000 +0900
> +++ ./streamlink/utils/parse.py	2022-04-01 21:48:47.000000000 +0900
> @@ -51,6 +51,8 @@ def parse_html(
>       - Removes XML declarations of invalid XHTML5 documents
>       - Wraps errors in custom exception with a snippet of the data in the message
>      """
> +    if isinstance(data, str):
> +        data = bytes(data, "utf8")
>      if isinstance(data, str) and data.lstrip().startswith("<?xml"):
>          data = re.sub(r"^\s*<\?xml.+?\?>", "", data)
>  
> ```
> 
> #4280 was also reported by a macOS user and closed as unconfirmed.

Thanks very much, thisisgk! You remind me that this issue did start to happen after I upgraded streamlink, I had not upgraded streamlink for a while before that, so the previous version that I was using should be 3.0.1 or lower.

**Comment by thisisgk** [comment]
> The issue only occurs when installing Streamlink via homebrew. When installing python3.10, creating a new virtualenv and installing streamlink and its dependencies via pip, it is working fine:
> 
> ```
> (venv) basti@bastis-iMac ~/D/streamlink (master)> streamlink -l debug 'https://ok.ru/video/3178148273793'
> [cli][debug] OS:         macOS 10.15.2
> [cli][debug] Python:     3.10.2
> [cli][debug] Streamlink: 3.2.0+9.g3762fe8
> ```

> The lxml version of Streamlink's homebrew formula is pinned to 4.8.0, but it's not using any pre-built wheels like when installing via pip. It instead relies on other system libraries, which means it's most likely an issue with homebrew's libxml2:

Please wait. This is nothing to do with Homebrew. I'm not using package managers for macOS (except pip).
I can see that lxml 4.8.0 ".so" files are linked to system libxml2 and libxslt from the output of `otool -L`.

And yes, it's working on macOS 10.15. I wrote:
> while it doesn't reproduce on macOS 10.15.7, it reproduces on 10.12.6 (both with Python 3.10.4 `venv`, then fresh `pip install streamlink`).

**Comment by bastimeyer** [comment]
> I'm not using package managers for macOS

I do in my macOS VM, and the homebrew package is broken because of its dynamically linked lxml dependency, whereas lxml installed from the **pypi wheel** is not, because it is statically linked and doesn't depend on any system libs:

```
$ otool -L /usr/local/Cellar/streamlink/3.2.0_1/libexec/lib/python3.10/site-packages/lxml/etree.cpython-310-darwin.so 
/usr/local/Cellar/streamlink/3.2.0_1/libexec/lib/python3.10/site-packages/lxml/etree.cpython-310-darwin.so:
	/usr/lib/libxslt.1.dylib (compatibility version 3.0.0, current version 3.26.0)
	/usr/lib/libexslt.0.dylib (compatibility version 9.0.0, current version 9.15.0)
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
 
$ otool -L /Users/basti/Desktop/streamlink/venv/lib/python3.10/site-packages/lxml/etree.cpython-310-darwin.so 
/Users/basti/Desktop/streamlink/venv/lib/python3.10/site-packages/lxml/etree.cpython-310-darwin.so:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```

> I can see that lxml 4.8.0 ".so" files are linked to system libxml2 and libxslt

This can of course also be a problem for non-homebrew installs if no lxml wheel is used.

**Comment by bastimeyer** [comment]
Building lxml manually with libxml2 installed from homebrew (2.9.13) works, btw:

```sh
$ /usr/local/bin/python3.10 -m virtualenv -p /usr/local/bin/python3.10 ./venv

$ brew install libxml2
$ export LDFLAGS="-L/usr/local/opt/libxml2/lib"
$ export CPPFLAGS="-I/usr/local/opt/libxml2/include"
$ export PKG_CONFIG_PATH="/usr/local/opt/libxml2/lib/pkgconfig"
$ pip install --no-binary=:all: lxml

$ pip install streamlink
$ streamlink 'https://ok.ru/video/3178148273793'
[cli][info] Found matching plugin okru for URL https://ok.ru/video/3178148273793
Available streams: 240p (worst), 360p, 480p, 720p (best)
```

**Comment by bastimeyer** [comment]
Short recap:

1. [at least lxml 4.6.4 is required since Streamlink 3.0.0](https://github.com/streamlink/streamlink/commit/6cad2faf1f58cd2d9f8c986814f2c10472c429b9)
2. [lxml 4.3.0 has set the minimum version requirement of libxml2 to 2.9.2](https://github.com/lxml/lxml/blob/53c5a224a4e6f8209a063ebc003cf296c5844b43/CHANGES.txt#L403-L404)
3. - [lxml 4.5.0+ wheels include libxml 2.9.10](https://github.com/lxml/lxml/blob/53c5a224a4e6f8209a063ebc003cf296c5844b43/CHANGES.txt#L209) (build logs expired)
   - [lxml 4.7.1+ wheels include libxml 2.9.12](https://github.com/lxml/lxml/blob/53c5a224a4e6f8209a063ebc003cf296c5844b43/CHANGES.txt#L56) (lxml 4.8.0 macOS 10.15 py310 wheel build log [here](https://github.com/lxml/lxml/runs/5231007381?check_suite_focus=true#step:6:5) - `STATIC_DEPS=true`)

**Installing wheels from pypi of any supported lxml version (4.6.4 - 4.8.0) works with Streamlink**, I've just tested it again.

----

macOS 10.15.2 (my VM) comes with libxml2 2.9.4, which should be supported since lxml 4.3.0.
The streamlink homebrew package also links against this version in its lxml build ([from the streamlink homebrew build log artifacts](https://github.com/Homebrew/homebrew-core/actions/runs/1940275039)):

```
Installing collected packages: lxml
  Running setup.py install for lxml: started
  Running command Running setup.py install for lxml
  Building lxml version 4.8.0.
  Building without Cython.
  Building against libxml2 2.9.4 and libxslt 1.1.29
  Building against libxml2/libxslt in the following directory: /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib
```

This old libxml2 version however causes issues with `lxml.etree.HTML()`, after Streamlink changed the input of this method in its 3.0.2 release via #4201 from `bytes` to `str`, because lxml otherwise can't correctly determine the encoding of HTML inputs when there's a `<meta charset="utf-8"/>` tag in the first X bytes of the input payload, which was the motivation for this change (see the added tests in the linked PR). This was noticed by me when Umlaut characters were not correctly decoded on certain sites.

If you revert this change while also using one of the libxml2 versions between 2.9.4 and 2.9.9 (not sure which one introduces the issue), eg. by using the dynamically linked system library, then `lxml.etree.HTML()` will still work. Whether it correctly handles the encoding of the example mentioned above, I can't tell.

**This is a bug in lxml which needs to be reported there.** The same input should work accross all supported versions, so it's not up to Streamlink to fix/change this.

----

You should however not use outdated versions of libxml2 which include various bugs and security issues:
https://gitlab.gnome.org/GNOME/libxml2/-/blob/master/NEWS

It's better to use the lxml wheels from pypi with static dependencies

```sh
$ pip install -U --binary-only=:all: lxml
```

or building lxml against the most recent libxml2 version, like I've shown above in my previous post. **This is also what the [streamlink homebrew package](https://github.com/Homebrew/homebrew-core/blob/2b070bc0188ea3b01aa226362a90f740f3d76983/Formula/streamlink.rb) should do**, so it might be worth reporting a bug there as well.

**Comment by thisisgk** [comment]
FYI.
I tried to see from which version of libxml2 works, and found that the self-built libxml2 2.9.4 just works. (force loaded dylib with DYLD_LIBRARY_PATH)

So macOS provided libxml2 2.9.4 (or libicucore) seems to have the problem. It's configured with `--with-iconv=no --with-icu=yes`, https://github.com/apple-oss-distributions/libxml2/blob/main/Makefile, whereas my self-built with `--with-iconv` (libxml2 default).

**Comment by bastimeyer** [comment]
The `streamlink` homebrew package has just been fixed (`3.2.0_2`), and `libxml2` is now a proper homebrew package dependency when building Streamlink's `lxml` python dependency. See Homebrew/homebrew-core#98468

The video URL posted in the OP has been taken down by the site since it was posted here, but this one for example now works:

```
$ streamlink -l debug 'https://ok.ru/video/3388934659879'
[cli][debug] OS:         macOS 10.15.2
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.2.0
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.3.1)
[cli][debug] Arguments:
[cli][debug]  url=https://ok.ru/video/3388934659879
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin okru for URL https://ok.ru/video/3388934659879
[utils.l10n][debug] Language code: en_US
Available streams: 144p (worst), 240p, 360p, 480p, 720p (best)
```

```
$ otool -L /usr/local/Cellar/streamlink/3.2.0_2/libexec/lib/python3.10/site-packages/lxml/etree.cpython-310-darwin.so
/usr/local/Cellar/streamlink/3.2.0_2/libexec/lib/python3.10/site-packages/lxml/etree.cpython-310-darwin.so:
        /usr/lib/libxslt.1.dylib (compatibility version 3.0.0, current version 3.26.0)
        /usr/lib/libexslt.0.dylib (compatibility version 9.0.0, current version 9.15.0)
        /usr/local/opt/libxml2/lib/libxml2.2.dylib (compatibility version 12.0.0, current version 12.13.0)
        /usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```

<details>
<summary>demonstration that it doesn't work with the system's libxml2</summary>

```
$ DYLD_LIBRARY_PATH=/usr/lib streamlink -l debug 'https://ok.ru/video/3388934659879'
[cli][debug] OS:         macOS 10.15.2
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.2.0
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.3.1)
[cli][debug] Arguments:
[cli][debug]  url=https://ok.ru/video/3388934659879
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin okru for URL https://ok.ru/video/3388934659879
[plugins.okru][error] Could not find metadata
error: No playable streams found on this URL: https://ok.ru/video/3388934659879
```
</details>

----

As said, users who don't use homebrew either have to install lxml's wheels from pypi or they have to avoid building lxml against the system's libxml2.

Closing...
